### DISCUSSION TITLE
Redis Bitmap 구조로 Redis 메모리 절약하기

### DISCUSSION THREAD

**Discussion by kmularise** [dailyge/dailyge-server#137] [Discussion]
## 📝 연관 PR

<br/><br/>

## ❓ 작성 이유
선착순 쿠폰 발급 참여 시, 서버 단에서 사용자의 이벤트 참여 여부를 Redis에 저장해두고 있었습니다. 이러한 사용자 이벤트 참여 여부를 저장할 수 있는 Redis 데이터 저장 방식에 대해 알아보고 bitmap을 선택한 이유를 정리하고자 이 글을 작성하게 되었습니다.

<br/><br/>

## bitmap
bitmap은 공간 비트(bit) 배열로 매핑된 비트 묶음을 말합니다. Redis에서 bitmap은 실제로는 데이터 타입은 아니며 String 타입이 bit 벡터 처럼 취급되어 bit 연산들로 정의되어 있습니다. String 타입은 최대 길이가 512 MB라서 최대 ![2^32](https://latex.codecogs.com/png.latex?2^{32})개의 bit를 설정할 수 있습니다.

<br/><br/> 

### 장점
- 비트 연산을 통해 setbit와 getbit 명령어를 O(1)로 실행할 수 있습니다. 
- 메모리를 절약할 수 있습니다.

<br/><br/>

### 사용 시 유의할 점
- 최대 offset에 대해서 고려해야 합니다.

### 활용 사례
- 사용자의 이벤트 참여 여부
- 이벤트 연속 참여 여부

<br/><br/>

## 📖 참고자료
- [Redis bitmaps](https://redis.io/docs/latest/develop/data-types/bitmaps/)

**Comment by devjun10** [comment]
좋네요. 어제 설명드린대로 bitmap을 사용하면 적은 공간으로 많은 데이터를 저장할 수 있습니다. 추가로 bitmap은 &, | 연산을 제공해서 이벤트 연속 참여 여부와 같은 연산도 쉽게 할 수 있으니 이 부분 한 번 학습해보셔요~

**Reply by kmularise** [reply]
좋아요! 말씀하신 부분도 학습해볼게요!
